### ISSUE TITLE
Laravel valet bind() to 0.0.0.0:443 failed (48: Address already in use)

### ISSUE THREAD

**Issue by eComEvo** [laravel/valet#707]
I have Laravel Valet 2.1.6 installed on Mac OS 10.14.2.

nginx 1.15.8 is installed using brew.

I restarted my Mac without installing any updates or new software, and now all `example.test` sites are giving a 502 error with the following showing in the `/usr/local/var/log/nginx/error.log` log:

    2019/01/17 20:38:47 [warn] 31277#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/etc/nginx/nginx.conf:1
    2019/01/17 20:38:47 [emerg] 31277#0: bind() to 0.0.0.0:443 failed (48: Address already in use)
    2019/01/17 20:38:47 [emerg] 31277#0: bind() to 0.0.0.0:443 failed (48: Address already in use)
    2019/01/17 20:38:47 [emerg] 31277#0: bind() to 0.0.0.0:443 failed (48: Address already in use)
    2019/01/17 20:38:47 [emerg] 31277#0: bind() to 0.0.0.0:443 failed (48: Address already in use)
    2019/01/17 20:38:47 [emerg] 31277#0: bind() to 0.0.0.0:443 failed (48: Address already in use)
    2019/01/17 20:38:47 [emerg] 31277#0: still could not bind()

At the same time, I get the following in the `/Users/Myself/.config/valet/Log/nginx-error.log` log:

    2019/01/17 20:41:34 [error] 32071#0: *1 upstream prematurely closed connection while reading response header from upstream, client: 127.0.0.1, server: vshred.test, request: "GET / HTTP/2.0", upstream: "fastcgi://unix:/Users/Myself/.config/valet/valet.sock:", host: "example.test"

When I run `ps ax -o pid,ppid,%cpu,vsz,wchan,command|egrep '(nginx|PID)'` I see this list:

      PID  PPID  %CPU      VSZ WCHAN  COMMAND
    32064     1   0.0  4306660 -      nginx: master process /usr/local/opt/nginx/bin/nginx -g daemon off;
    32065 32064   0.0  4333284 -      nginx: worker process
    32066 32064   0.0  4332260 -      nginx: worker process
    32067 32064   0.0  4333284 -      nginx: worker process
    32068 32064   0.0  4333284 -      nginx: worker process
    32069 32064   0.0  4326116 -      nginx: worker process
    32070 32064   0.0  4316900 -      nginx: worker process
    32071 32064   0.0  4368236 -      nginx: worker process
    32072 32064   0.0  4331236 -      nginx: worker process
    32073 32064   0.0  4326116 -      nginx: worker process
    32074 32064   0.0  4340452 -      nginx: worker process
    32075 32064   0.0  4333284 -      nginx: worker process
    32076 32064   0.0  4334308 -      nginx: worker process
    36815  1406   0.0  4268060 -      egrep (nginx|PID)

None of the following solves the issue:

* `sudo killall nginx`
* `brew services restart nginx`
* `brew services restart php`
* `valet restart`
* Restarting my Mac
* `valet uninstall && valet install` then `valet park` on the relevant dir

Apache is not running as a conflicting service.

I tried doing `sudo /usr/local/opt/nginx/bin/nginx -g 'daemon off;'` and got this:

    nginx: [emerg] bind() to 0.0.0.0:80 failed (48: Address already in use)
    nginx: [emerg] bind() to 0.0.0.0:443 failed (48: Address already in use)
    nginx: [emerg] bind() to 0.0.0.0:60 failed (48: Address already in use)
    nginx: [emerg] bind() to 0.0.0.0:80 failed (48: Address already in use)
    nginx: [emerg] bind() to 0.0.0.0:443 failed (48: Address already in use)
    nginx: [emerg] bind() to 0.0.0.0:60 failed (48: Address already in use)
    nginx: [emerg] bind() to 0.0.0.0:80 failed (48: Address already in use)
    nginx: [emerg] bind() to 0.0.0.0:443 failed (48: Address already in use)
    nginx: [emerg] bind() to 0.0.0.0:60 failed (48: Address already in use)
    nginx: [emerg] bind() to 0.0.0.0:80 failed (48: Address already in use)
    nginx: [emerg] bind() to 0.0.0.0:443 failed (48: Address already in use)
    nginx: [emerg] bind() to 0.0.0.0:60 failed (48: Address already in use)
    nginx: [emerg] bind() to 0.0.0.0:80 failed (48: Address already in use)
    nginx: [emerg] bind() to 0.0.0.0:443 failed (48: Address already in use)
    nginx: [emerg] bind() to 0.0.0.0:60 failed (48: Address already in use)
    nginx: [emerg] still could not bind()

Running `sudo lsof -i tcp:80` produces:

    COMMAND   PID USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
    nginx   42220 root    7u  IPv4 0x7ac8eae7874ccb11      0t0  TCP *:http (LISTEN)
    nginx   42221 Myself    7u  IPv4 0x7ac8eae7874ccb11      0t0  TCP *:http (LISTEN)
    nginx   42222 Myself    7u  IPv4 0x7ac8eae7874ccb11      0t0  TCP *:http (LISTEN)
    nginx   42223 Myself    7u  IPv4 0x7ac8eae7874ccb11      0t0  TCP *:http (LISTEN)
    nginx   42224 Myself    7u  IPv4 0x7ac8eae7874ccb11      0t0  TCP *:http (LISTEN)
    nginx   42225 Myself    7u  IPv4 0x7ac8eae7874ccb11      0t0  TCP *:http (LISTEN)
    nginx   42226 Myself    7u  IPv4 0x7ac8eae7874ccb11      0t0  TCP *:http (LISTEN)
    nginx   42227 Myself    7u  IPv4 0x7ac8eae7874ccb11      0t0  TCP *:http (LISTEN)
    nginx   42228 Myself    7u  IPv4 0x7ac8eae7874ccb11      0t0  TCP *:http (LISTEN)
    nginx   42229 Myself    7u  IPv4 0x7ac8eae7874ccb11      0t0  TCP *:http (LISTEN)
    nginx   42230 Myself    7u  IPv4 0x7ac8eae7874ccb11      0t0  TCP *:http (LISTEN)
    nginx   42231 Myself    7u  IPv4 0x7ac8eae7874ccb11      0t0  TCP *:http (LISTEN)
    nginx   42232 Myself    7u  IPv4 0x7ac8eae7874ccb11      0t0  TCP *:http (LISTEN)

Basically the same output when I run that command for port 443.

This valet issue post suggests `valet domain test` might fix it, but that didn't help:
https://github.com/laravel/valet/issues/488

Any idea where to go from here?

**Comment by drbyte** [comment]
> `the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/etc/nginx/nginx.conf:1`

When you run `brew services list`, is nginx running as root?  If not, was that intentional on your part?

(Valet normally runs nginx as root, and IIRC a default homebrew install of nginx will also have it run as root)
(starting nginx with sudo will make it run as root, which is what valet does: `sudo brew services start nginx`)

**Comment by eComEvo** [comment]
Not starting nginx as root was not intentional. I didn't change anything related to that.

I tried `sudo brew services stop nginx` followed by `sudo brew services start nginx` and the 502 error continues. Also tried `sudo valet restart` but no luck either.

All that happened when I ran those commands is that now the `the "user" directive makes sense only if the master process` error is missing from the error log but the other error lines are the same.

**Comment by drbyte** [comment]
## For reference, compared to a working installation:
When I have nginx already running with Valet and run `sudo /usr/local/opt/nginx/bin/nginx -g 'daemon off;'` I get the same `bind() to 0.0.0.0:80 failed (48: Address already in use)` responses (all 3 ports) as you posted in your first message. This is expected.
When I `sudo brew services stop nginx` and then run `sudo /usr/local/opt/nginx/bin/nginx -g 'daemon off;'`, it gives no output because it's running without problem. As expected.
Canceling that with CTRL+C and restarting nginx with homebrew also works correctly again. As expected.

Running the other commands you mentioned gives the same output as you (granted, different PIDs and DEVICE):
- `sudo lsof -i tcp:80`
- `ps ax -o pid,ppid,%cpu,vsz,wchan,command|egrep '(nginx|PID)'`

## things to check
Do the following give you the same output?

Asking nginx to test my configs gives:
```sh
sudo /usr/local/opt/nginx/bin/nginx -t
```
> nginx: the configuration file /usr/local/etc/nginx/nginx.conf syntax is ok
> nginx: configuration file /usr/local/etc/nginx/nginx.conf test is successful

Do you have any extra nginx `.conf` files?? Do any of them differ from the default stubs that valet uses (from `~/.composer/vendor/laravel/valet/cli/stubs`)?

```sh
ls -al /usr/local/etc/nginx/
```
```
total 136
drwxr-xr-x  18 Myself  admin   576 Jan  8 18:48 .
drwxrwxr-x  30 Myself  admin   960 Jan 15 11:48 ..
-rw-r--r--   1 Myself  admin  1077 Dec 27 18:58 fastcgi.conf
-rw-r--r--   1 Myself  admin  1077 Dec 27 18:58 fastcgi.conf.default
-rw-r--r--@  1 Myself  admin   924 Jan  8 18:47 fastcgi_params
-rw-r--r--   1 Myself  admin  1007 Dec 27 18:58 fastcgi_params.default
-rw-r--r--   1 Myself  admin  2837 Dec 27 18:58 koi-utf
-rw-r--r--   1 Myself  admin  2223 Dec 27 18:58 koi-win
-rw-r--r--   1 Myself  admin  5231 Dec 27 18:58 mime.types
-rw-r--r--   1 Myself  admin  5231 Dec 27 18:58 mime.types.default
-rw-r--r--   1 Myself  admin   821 Jan  4 21:35 nginx.conf
-rw-r--r--   1 Myself  admin  2680 Dec 27 18:58 nginx.conf.default
-rw-r--r--   1 Myself  admin   636 Dec 27 18:58 scgi_params
-rw-r--r--   1 Myself  admin   636 Dec 27 18:58 scgi_params.default
-rw-r--r--   1 Myself  admin   664 Dec 27 18:58 uwsgi_params
-rw-r--r--   1 Myself  admin   664 Dec 27 18:58 uwsgi_params.default
drwxr-xr-x   3 root   admin    96 Nov 27 23:31 valet
-rw-r--r--   1 Myself  admin  3610 Dec 27 18:58 win-utf
```

```sh
ls -al /usr/local/etc/nginx/valet
```
```
total 8
drwxr-xr-x   3 root   admin   96 Nov 27 23:31 .
drwxr-xr-x  18 Myself  admin  576 Jan  8 18:48 ..
-rw-r--r--@  1 Myself  admin  901 Jan  4 21:35 valet.conf
```

Have you added (or have any broken) Nginx configs in your `~/.config/valet/Nginx` or `~/.config/valet/Certificates` directories? 

Try renaming your `~/.config/valet` to valet-backup and then run `valet install`. Does this resolve the problem? If so, something's amuck in your site definitions.

**Comment by drbyte** [comment]
While your report of `lsof` suggests nothing else competing on port 80, for thoroughness it may also be worth expanding the `ps` check to look for apache (or httpd alternative), in case those have gotten started on your machine somehow:

`ps ax -o pid,ppid,%cpu,vsz,wchan,command|egrep '(nginx|PID|httpd|apache)'`

And in theory the following should say file not found if apache has never been run, but if it has, then maybe it's running now? Its other log files would reveal more about it.
`tail -5 /var/log/httpd/error_log`

**Comment by eComEvo** [comment]
`ps ax -o pid,ppid,%cpu,vsz,wchan,command|egrep '(nginx|PID|httpd|apache)'` gives no results except for nginx in this list, unless I do `sudo brew services stop nginx` in which case nothing shows but my recent search in the list.

`tail -5 /var/log/httpd/error_log` gives `tail: cannot open '/var/log/httpd/error_log' for reading: No such file or directory`

I get the same output as you for `sudo /usr/local/opt/nginx/bin/nginx -t`.

`ls -al /usr/local/etc/nginx/` shows exactly the same files as you have in your list, with this one strange exception where `wheel` is the owner of the valet dir instead of `root`:
```
drwxr-xr-x  3 Myself wheel   96 Jan 21 17:36 valet
```

Completely deleting all valet config/cert directories, uninstalling valet again, then reinstalling did not solve it.

**Comment by eComEvo** [comment]
Confirmed....it was PHP 7.3. After removing 7.3 and `brew install php@7.2`, then reinstalling the correct version of xdebug for 7.2, everything works fine.
